File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int>result;
        for(auto as:asteroids)
        {
            bool isAdd=true;
            while(!result.empty()&&result.back()>0&&as<0)
            {
                if(result.back()==abs(as))
                {
                    isAdd=false;
                    result.pop_back();   
                    break;
                }
                else if(result.back()<abs(as))
                {
                    result.pop_back();
                }
                else
                {
                    isAdd=false;
                    break;
                }
            }
            if(isAdd)
                result.push_back(as);
        }
        return result;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> st = new Stack<>();
        for(int asteroid:asteroids)
        {
            if(asteroid<0)
            {
                while(st.size() > 0 && st.peek() > 0 && st.peek()+asteroid<0)
                {
                    st.pop();
                }
                if (st.size() > 0 && st.peek() + asteroid == 0)
                {
                    st.pop();
                }
                else if(st.isEmpty() || st.peek() < 0)
                    st.push(asteroid);
            }
            else
            {
                st.push(asteroid);
            }
        }

        return st.stream().mapToInt(Integer::intValue).toArray();
    }
}